src/Symfony/Component/VarDumper/Caster/Caster.php

```diff
@@ -0,0 +1,106 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\VarDumper\Caster;
+
+/**
+ * Helper for filtering out properties in casters.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class Caster
+{
+    const EXCLUDE_VERBOSE = 1;
+    const EXCLUDE_VIRTUAL = 2;
+    const EXCLUDE_DYNAMIC = 4;
+    const EXCLUDE_PUBLIC = 8;
+    const EXCLUDE_PROTECTED = 16;
+    const EXCLUDE_PRIVATE = 32;
+    const EXCLUDE_NULL = 64;
+    const EXCLUDE_EMPTY = 128;
+    const EXCLUDE_NOT_IMPORTANT = 256;
+
+    const PREFIX_VIRTUAL = "\0~\0";
+    const PREFIX_DYNAMIC = "\0+\0";
+    const PREFIX_PROTECTED = "\0*\0";
+
+    /**
+     * Casts objects to arrays and adds the dynamic property prefix.
+     *
+     * @param object          $obj       The object to cast.
+     * @param ReflectionClass $reflector The class reflector to use for inspecting the object definition.
+     *
+     * @return array The array-cast of the object, with prefixed dynamic properties.
+     */
+    public static function castObject($obj, \ReflectionClass $reflector)
+    {
+        if ($reflector->hasMethod('__debugInfo')) {
+            $v = $obj->__debugInfo();
+        } else {
+            $v = (array) $obj;
+        }
+
+        $a = array();
+
+        foreach ($v as $k => $v) {
+            if (!isset($k[0]) || ("\0" !== $k[0] && !$reflector->hasProperty($k))) {
+                $a[self::PREFIX_DYNAMIC.$k] = $v;
+            } else {
+                $a[$k] = $v;
+            }
+        }
+
+        return $a;
+    }
+
+    /**
+     * Filters out the specified properties.
+     *
+     * @param array $a                The array containing the properties to filter.
+     * @param int   $filter           A bit field of Caster::EXCLUDE_* constants specifying which properties to filter out.
+     * @param array $listedProperties List of properties to exclude when Caster::EXCLUDE_VERBOSE is set, and to preserve when Caster::EXCLUDE_NOT_IMPORTANT is set.
+     *
+     * @return array The filtered array
+     */
+    public static function filter(array $a, $filter, array $listedProperties = array())
+    {
+        if (self::EXCLUDE_VERBOSE & $filter) {
+            foreach ($listedProperties as $k) {
+                unset($a[$k]);
+            }
+        }
+
+        foreach ($a as $k => $v) {
+            $type = self::EXCLUDE_PUBLIC;
+
+            switch (true) {
+                case (self::EXCLUDE_NULL & $filter) && null === $v:
+                case (self::EXCLUDE_EMPTY & $filter) && empty($v):
+                case (self::EXCLUDE_NOT_IMPORTANT & $filter) && !in_array($k, $listedProperties, true):
+                    $type = $filter;
+                    break;
+                case isset($k[1]) && "\0" === $k[0]:
+                    switch ($k[1]) {
+                        case '~': $type = self::EXCLUDE_VIRTUAL; break;
```

Comment:
I would make this multiline to make it more readable (seeing easily that we have some login in the case, not only a condition falling through to the next case).
currently, the code here is very similar to the code on lines 85 and 86 at first look, while the structure of the code is actually totally different